### ISSUE TITLE
Symbolic observation drops visible mobs: unclamped write in _add_mob_to_map

### ISSUE THREAD

**Issue by pulkit6732** [MichaelTMatthews/Craftax#57]
# Symbolic observation drops visible mobs: unclamped write in `_add_mob_to_map`

## Summary

In the symbolic renderers, the mob channels are written with an unclamped, unconditional
`.set()`. Every mob slot performs a write on every step, including mobs that are off screen
and mobs whose `mask` is 0. Those writes store `0`, but not always into their own cell, so a
mob standing next to the agent can be erased from the observation.

Measured against `Craftax-Classic-Symbolic-v1` on `craftax==1.6.1`: **87 of 25,600
observations (0.340%) disagree with the environment state**, with 87 real mobs missing.
The defect only ever deletes mobs; it never invents them.

## Reproduction

`repro_mob_map.py` (attached) rolls out the real environment and compares the mob channels of
the returned observation against a reference built directly from `state`. It needs nothing
beyond `craftax`, `jax` and `numpy`.

```
python repro_mob_map.py
```

Output on 1.6.1:

```
environment          : Craftax-Classic-Symbolic-v1
observations checked : 25,600 (64 envs x 400 steps)
mismatched           : 87 (0.340%)
mobs missing         : 87
mobs phantom         : 0

FAIL: observation does not match environment state
```

The script exits non-zero when the observation disagrees with the state, so it also works as a
regression test.

## Root cause

`craftax/craftax_classic/renderer.py`, `_add_mob_to_map`:

```python
on_screen = jnp.logical_and(
    local_position >= 0, local_position < jnp.array([OBS_DIM[0], OBS_DIM[1]])
).all()
on_screen *= mobs.mask[mob_index]

mob_map = mob_map.at[local_position[0], local_position[1], mob_type_index].set(
    on_screen.astype(jnp.uint8)
)
```

`on_screen` is computed correctly, but it is used only as the *value*. It never gates whether
the write happens, and `local_position` is never clamped before being used as an index. That
produces two distinct failure modes:

**A. Negative index wraps.** JAX wraps negative indices rather than dropping them:

```python
base = jnp.zeros((7, 9, 4), jnp.uint8).at[3, 4, 0].set(1)
base.at[-6, 4, 0].set(0)[3, 4, 0]   # -> 0, the mob at row 3 was cleared
base.at[20, 4, 0].set(0)[3, 4, 0]   # ->  1, out-of-range positive is dropped
```

With `OBS_DIM = (7, 9)`, a mob four tiles above the viewport has `local_position[0] == -4`,
which wraps to row 3 — the centre row, where the agent stands. Only mobs above or to the left
of the view are destructive.

**B. Inactive mob in bounds.** A mob with `mask == 0` whose stale position happens to fall
inside the view writes `0` into that cell, clearing whatever live mob occupies it. No
out-of-range index is involved, so clamping alone does not fix this case.

Across a 200,000-state sweep the two modes account for 145 and 24 states respectively. Of 169
destructive-write opportunities, 140 persisted into the final observation; the remainder were
repaired by accident when the live mob happened to be written later in the scan.

## Scope

| File | Line | Function |
| --- | --- | --- |
| `craftax/craftax_classic/renderer.py` | 56 | `render_craftax_symbolic` (classic) |
| `craftax/craftax/renderer.py` | 56 | `render_craftax_symbolic` |
| `craftax/craftax/renderer.py` | 963 | `render_craftax_text` |

Full Craftax carries a larger view, more mob slots, and a channel index that is itself
data-dependent (`mob_class_index * mob_types_per_class + mobs.type_id[mob_index]`), so an
inactive slot contributes a stale channel as well as a stale position. The 0.340% figure was
measured on `craftax_classic` only.

The pixel renderers compute the same unclamped `local_position` but composite with alpha
weighted by `on_screen`, so an off-screen mob contributes nothing rather than clearing a cell.
They are not affected.

## Suggested fix

Send non-visible writes off-grid and let `mode="drop"` discard them, so only genuinely visible
mobs touch the map. This handles both failure modes and avoids reading the array back:

```python
write_position = jnp.where(
    on_screen, local_position, jnp.array([OBS_DIM[0], OBS_DIM[1]])
)
mob_map = mob_map.at[write_position[0], write_position[1], mob_type_index].set(
    on_screen.astype(jnp.uint8), mode="drop"
)
```

A patch for all three sites is attached.

I also tried `jnp.clip` combined with `jnp.maximum`. It is correct, but the extra gather inside
the scan cost noticeably more throughput than the version above, so I discarded it. Rough CPU
figures at 256 envs, best of 5 runs of 200 steps: 2,250 env-steps/s upstream, 1,969 with the
proposed fix, 1,671 with clip + maximum. These are indicative only — single machine, CPU
backend, and the run-to-run spread was several percent. Worth confirming on the accelerator you
care about.

## Note on existing baselines

This changes observations for affected states, which is the intent, but published results were
produced against the current encoding. It may deserve a release note, and whether it should be
gated for reproducibility is your call.

## Environment

```
craftax  1.6.1
jax      0.11.0
jaxlib   0.11.0
numpy    2.5.2
python   3.14.7
backend  cpu
```

[repro_mob_map.py](https://github.com/user-attachments/files/31922244/repro_mob_map.py)
